### ISSUE TITLE
TS declaration error: Type 'IndexNames<DBTypes, StoreName>' does not satisfy the constraint 'string'

### ISSUE THREAD

**Issue by sumbricht** [jakearchibald/idb#311]
When compiling any project that relies on the idb library, I'm getting the following TypeScript error:
`Error: node_modules/idb/build/entry.d.ts:359:45 - error TS2344: Type 'IndexNames<DBTypes, StoreName>' does not satisfy the constraint 'string'`

Looking at entry.d.ts, this error is actually correct.
The error occurs in the property definition of `indexNames`:
```ts
export interface IDBPObjectStore<DBTypes extends DBSchema | unknown = unknown, StoreName extends StoreNames<DBTypes> = StoreNames<DBTypes>, Mode extends IDBTransactionMode = 'readonly'> extends IDBPObjectStoreExtends {
    readonly indexNames: TypedDOMStringList<IndexNames<DBTypes, StoreName>>;
    [..]
}
```

`TypesDOMStringList<T extends string>`expects `IndexedNames` to be of assignable to `string` but this turns out to be `string | number` due to the way `IndexedNames` is defined. The expression `keyof DBTypes[StoreName]['indexes']` is unfortunately `string | number` even if it should intuitively be `string` only.

To verify this, let's look at a simplified example:
```ts
interface KeyValue {
    [s: string]: any;
}
type Key = keyof KeyValue // you'd expect `string` here, but it's `string | number`
```

To solve the problem, I propose changing the type `IndexNames` from
`export declare type IndexNames<DBTypes extends DBSchema | unknown, StoreName extends StoreNames<DBTypes>> = DBTypes extends DBSchema ? keyof DBTypes[StoreName]['indexes'] : string;`
to
`export declare type IndexNames<DBTypes extends DBSchema | unknown, StoreName extends StoreNames<DBTypes>> = DBTypes extends DBSchema ? keyof DBTypes[StoreName]['indexes'] & string : string;`

As it's a very quick fix it may be easier for you to just do it; if you'd like a PR, just let me know :-).

Thanks a lot for your help and the great work on this library!

**Comment by caboodal** [comment]
I'm also getting this issue but only after upgrading typescript to 5.4.5: -

X [ERROR] TS2344: Type 'IndexNames<DBTypes, StoreName>' does not satisfy the constraint 'string'.
  Type 'string | number' is not assignable to type 'string'.
    Type 'number' is not assignable to type 'string'. [plugin angular-compiler]

    node_modules/idb/build/entry.d.ts:359:44:
      359 │ ...y indexNames: TypedDOMStringList<IndexNames<DBTypes, StoreName>>;

**Comment by philkunz** [comment]
+1

**Comment by simondingle** [comment]
+1

**Comment by philkunz** [comment]
"@tempfix/idb": "^8.0.3" solves the problem for now by implementing the fix suggested above.

**Comment by leesprojects** [comment]
+1

**Comment by shruggr** [comment]
+1

**Comment by ksmit** [comment]
+1

**Comment by n0daft** [comment]
+1

**Comment by tambor81** [comment]
+1

**Comment by tambor81** [comment]
> As it's a very quick fix it may be easier for you to just do it; if you'd like a PR, just let me know :-).

Would you be so kind to make the PR?
 Now we are upgrading to the latest Angular version17 and 18 and also upgrading typescript this compiling error you found is appearing,  it'd be awesome if you could get the credit for your solution.

**Comment by y-reut** [comment]
+1

**Comment by AlexandraGae** [comment]
+1

**Comment by elisheva-vol** [comment]
+1

**Comment by uaKorona** [comment]
+1

**Comment by 5Gears0Chill** [comment]
+1

**Comment by elisheva-vol** [comment]
any update?

**Comment by oliviercochet** [comment]
Maybe someone may create the PR in tagging the issuer for credit?

**Comment by shruggr** [comment]
I tried to do that today, but only the dust folder is available from the npm package, so I don’t have the code for the fix.

**Comment by uaKorona** [comment]
Any updates ? I'm also blocked with angular 18v updating ...

**Comment by uaKorona** [comment]
I asked Copilot - is it possible to solve this problem locally ? 

It proposed to override typing of the idb lib locally 

1. Add into tsconfig a reference to local types
```
"typeRoots": ["./custom-types", "node_modules/@types"]
```

2. Redefine types in ./custom-types/my-idb.d.ts (for example) 
```
// Declare the module to augment it
declare module 'idb' {

  export type IndexNames<
    DBTypes extends DBSchema | unknown,
    StoreName extends StoreNames<DBTypes>,
  > = DBTypes extends DBSchema ? keyof DBTypes[StoreName]['indexes'] : string | number;
}
```

3. Restart your typescript server. Joy ) 

However, this approach didn't work for me )). Maybe I'm doing something wrong, or perhaps locally overriding typings isn't possible at all?

**Comment by aaronshaf** [comment]
@uaKorona I don't see the PR for this, did you intend that with the branch?

**Comment by uaKorona** [comment]
> @uaKorona I don't see the PR for this, did you intend that with the branch?

No, I’ve made a personal fork with a minimal fix and am waiting for the official resolution.

**Comment by sumbricht** [comment]
FYI, as the original poster I considered creating a PR, but my confidence that it'll get merged is a bit on the low side as there hasn't been a commit since Dec 1, 2023.

Anyway, suffering with this issue I got to know a very handy package that I'd like to share: [patch-package](https://github.com/ds300/patch-package). Just apply any hotfix directly in the node_modules folder and then run patch-package to create a patch file:

`npx patch-package idb`

This will create a file `patches/idb+8.0.0.patch` with the following content:
```
diff --git a/node_modules/idb/build/entry.d.ts b/node_modules/idb/build/entry.d.ts
index febdcf4..7651ec4 100644
--- a/node_modules/idb/build/entry.d.ts
+++ b/node_modules/idb/build/entry.d.ts
@@ -103,7 +103,7 @@ export declare type StoreKey<DBTypes extends DBSchema | unknown, StoreName exten
  * @template DBTypes DB schema type, or unknown if the DB isn't typed.
  * @template StoreName Names of the object stores to get the types of.
  */
-export declare type IndexNames<DBTypes extends DBSchema | unknown, StoreName extends StoreNames<DBTypes>> = DBTypes extends DBSchema ? keyof DBTypes[StoreName]['indexes'] : string;
+export declare type IndexNames<DBTypes extends DBSchema | unknown, StoreName extends StoreNames<DBTypes>> = DBTypes extends DBSchema ? keyof DBTypes[StoreName]['indexes'] & string : string;
 /**
  * Extract the types of indexes in certain object stores from the DB schema type.
  *
```

To apply the fixes in all your patch files (e.g. after a fresh install of node_modules), either manually run `npx patch-package` or put it into a postinstall script.

Hope this helps mitigate the pain / include such fixing into your CI pipeline ;-).

**Comment by aaronshaf** [comment]
@uaKorona if you could create a PR with your commit, that'd be great!

**Comment by philkunz** [comment]
@jakearchibald Any chance you could take a look at this, so we can deprecate the @tempfix/idb package?

**Comment by dylanvdmerwe** [comment]
Any feedback on this blocking issue?

**Comment by sumbricht** [comment]
Ok, I now created PR #336 for this. @jakearchibald, could you please have a look & merge?

**Comment by jakearchibald** [comment]
Thanks all! Sorry it's taken so long to get to this. I'll try to get through the issues and PRs over the next week.

**Comment by jakearchibald** [comment]
I've also changed the build of this project to catch issues like this in future. I was using a hacky typscript build because when I started this project the official plugin was trash. That's no longer the case.
